core/templates/dev/head/pages/topic_editor/subtopics_editor/SubtopicsListTabDirective.js

```diff
@@ -34,90 +34,155 @@ oppia.directive('subtopicsListTab', [
-              $scope.subtopics.indexOf(subtopic));
-          };
+                  $scope.openSubtopicTitleEditor = function() {
+                    $scope.subtopicTitleEditorIsShown = true;
+                  };
 
-          $scope.setSubtopic = function(subtopic) {
-            $scope.subtopicIndexToDisplay = $scope.subtopics.indexOf(subtopic);
-            $scope.editableTitle = subtopic.getTitle();
-            TopicEditorStateService.loadSubtopicPage(
-              $scope.topic.getId(), subtopic.getId());
-          };
+                  $scope.closeSubtopicTitleEditor = function() {
+                    $scope.subtopicTitleEditorIsShown = false;
+                  };
 
-          $scope.$on(EVENT_SUBTOPIC_PAGE_LOADED, function() {
-            $scope.subtopicPage = TopicEditorStateService.getSubtopicPage();
-            $scope.subtopicDisplayed = true;
-            $scope.subtopicEditorIsShown = false;
-            $scope.htmlData = $scope.subtopicPage.getHtmlData();
-          });
+                  $scope.openPreviewSubtopicPage = function(htmlData) {
+                    $scope.subtopicEditorIsShown = false;
+                    $scope.htmlData = htmlData;
+                  };
 
-          $scope.openSubtopicTitleEditor = function() {
-            $scope.subtopicTitleEditorIsShown = true;
+                  $scope.closePreviewSubtopicPage = function(previewHtmlData) {
+                    $scope.subtopicEditorIsShown = true;
+                    $scope.htmlData = previewHtmlData;
+                  };
+
+                  $scope.save = function() {
+                    $uibModalInstance.close({
+                      newTitle: $scope.editableTitle,
+                      newHtmlData: $scope.htmlData
+                    });
+                  };
+
+                  $scope.cancel = function() {
+                    $uibModalInstance.dismiss('cancel');
+                  };
+                }
+              ]
+            });
+
+            modalInstance.result.then(function(newValues) {
+              $scope.subtopicPage = TopicEditorStateService.getSubtopicPage();
+              var newTitle = newValues.newTitle;
+              var newHtmlData = newValues.newHtmlData;
+
+              if ((newTitle !== subtopic.getTitle()) && (newTitle !== '')) {
+                TopicUpdateService.setSubtopicTitle(
+                  $scope.topic, subtopic.getId(), newTitle);
+              }
+              if (newHtmlData !== $scope.subtopicPage.getHtmlData()) {
+                TopicUpdateService.setSubtopicPageHtmlData(
+                  $scope.subtopicPage, subtopic.getId(), newHtmlData);
+                TopicEditorStateService.setSubtopicPage($scope.subtopicPage);
+              }
+            });
           };
 
-          $scope.closeSubtopicTitleEditor = function() {
-            $scope.subtopicTitleEditorIsShown = false;
+          $scope.getSkillEditorUrl = function(subtopicIndex, skillIndex) {
```

Comment:
This seems a bit oddly factored. I would expect this method to take a skillId or a skill object and construct the URL, and not have to figure out the indexing as well.

Reply:
Done, changed it to use just skillId.